### PR TITLE
prevent unpredictable results with network create|remove

### PR DESCRIPTION
**Pull request by baude**
due to a lack of "locking" on cni operations, we could get ourselves in trouble when doing rapid creation or removal of networks.  added a simple file lock to deal with the collision and because it is not considered a performent path, use of the file lock should be ok.  if proven otherwise in the future, some generic shared memory lock should be implemented for libpod and also used here.

Fixes: #7807

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: pkg/api/handlers/compat/networks.go [hunk 1]
```diff
@@ -305,8 +306,9 @@ func RemoveNetwork(w http.ResponseWriter, r *http.Request) {
 		utils.Error(w, "network not found", http.StatusNotFound, define.ErrNoSuchNetwork)
 		return
 	}
-	if err := network.RemoveNetwork(config, name); err != nil {
+	if err := network.RemoveNetwork(config, name, config.Engine.TmpDir); err != nil {
```
Comment (rhatdan) [inline pkg/api/handlers/compat/networks.go:309]: Does this path make sense?  Should this be the cni path?  IE Could multiple containers create content by using different TmpDirs?

Comment (baude) [reply pkg/api/handlers/compat/networks.go:309]: this is using libpod's tmpdir to get a lock for all of cni.  at this point, containers dont matter.  the cni path is for where the plugins live on the filesystem so we know how/where to call them as they are not part of path.  the lock in question is "global, i need cni to not change for a second" lock.

File: pkg/domain/infra/abi/network.go [hunk 2]
```diff
@@ -115,6 +116,14 @@ func (ic *ContainerEngine) NetworkCreate(ctx context.Context, name string, optio
 		err      error
 		fileName string
 	)
+	config, err := ic.Libpod.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	// Acquire a lock for CNI
+	if _, err := storage.GetLockfile(filepath.Join(config.Engine.TmpDir, network.LockFileName)); err != nil {
```
Comment (vrothberg) [inline pkg/domain/infra/abi/network.go:124]: This does not yet acquire the lock.

`storage.GetLockfile` returns a `Locker, error` and we currently ignore the `Locker`.

To acquire the lock, we need to:
```golang
lock, err := storage.GetLockfile(...)
...
lock.Lock()
defer lock.Unlock()
```

File: pkg/network/network.go [hunk 3]
```diff
@@ -167,7 +169,11 @@ func ValidateUserNetworkIsAvailable(config *config.Config, userNet *net.IPNet) e
 
 // RemoveNetwork removes a given network by name.  If the network has container associated with it, that
 // must be handled outside the context of this.
-func RemoveNetwork(config *config.Config, name string) error {
+func RemoveNetwork(config *config.Config, name string, lockPath string) error {
+	// Acquire a lock for network removal
+	if _, err := storage.GetLockfile(filepath.Join(lockPath, LockFileName)); err != nil {
+		return err
+	}
```
Comment (vrothberg) [inline pkg/network/network.go:176]: Same issue as above.

I think it's _very_ risky to use the lock here _and_ outside of the package. It suggests that code outside of the package (i.e., `createMacVLAN` and `createBridge`) that needs the lock should rather live inside the package. Callers shouldn't need to worry about concurrency issues; the risk of missing to lock or running into deadlocks is high.

Comment (baude) [reply pkg/network/network.go:176]: yeah, i waffled on this as well.  I'm not sure I would categorize the risk similar but it is a fair comment.  I'll consider a move.

**Review by vrothberg — COMMENTED** [review]
LGTM, nice work!

**Comment by baude** [comment]
merge me

**Pull request merged**